backends/nxp/backend/ir/tflite_optimizer/optimizer.py

```diff
@@ -0,0 +1,229 @@
+                builder, conversion_config
+            ),
+            Optimization.FUSE_CAST_OPERATORS: FuseCastOperators(
+                builder, conversion_config
+            ),
+            Optimization.REMOVE_CAST_OPERATORS_WITH_NO_EFFECT: RemoveCastOperatorsWithNoEffect(
+                builder, conversion_config
+            ),
+            Optimization.MOVE_ACTIVATION_BEFORE_CONCAT: MoveActivationBeforeConcatenation(
+                builder, conversion_config
+            ),
+            Optimization.COMBINE_HARD_SIGMOID_AND_MUL_INTO_HARD_SWISH: CombineHardSigmoidAndMulIntoHardSwish(
+                builder, conversion_config
+            ),
+            Optimization.REPLACE_AVERAGE_POOL_BEFORE_FULLY_CONNECTED_WITH_SUM: ReplaceAveragePoolBeforeFullyConnectedWithSum(
+                builder, conversion_config
+            ),
+        }
+
+    def optimize(
+        self,
+        optimization_whitelist: list[Optimization] | None = None,
+        optimization_blacklist: list[Optimization] | None = None,
+    ):
+        """Apply optimizations to the TFLite model encapsulated by 'self._builder'.
+        :param optimization_whitelist: A list of optimizations to apply to the model.
+        :param optimization_blacklist: A list of optimizations to NOT apply to the model.
+
+        At least one of 'optimization_whitelist' and 'optimization_blacklist' must be 'None'.
+        If both are 'None', all optimizations are applied.
+
+        The optimizations will be applied multiple times in a loop, until the model is fully optimized.
+        """
+
+        if optimization_whitelist is not None and optimization_blacklist is not None:
+            logger.e(
+                logger.Code.INVALID_OPTIMIZATION,
+                "Optimization whitelist and blacklist cannot both be specified.",
+            )
+
+        if optimization_whitelist is not None:
+            optimizations = optimization_whitelist
+        else:
+            # Apply all optimizations
+            optimizations = list(Optimization)
+
+        if optimization_blacklist is not None:
+            for o in optimization_blacklist:
+                try:
+                    optimizations.remove(o)
+                except ValueError:
+                    logger.w(
+                        f"Optimization blacklist contains invalid optimization '{o}'."
+                    )
+
+        # Execute the optimizations until the model is fully optimized.
+        for _i in range(self.optimization_application_limit):
+            run_again = False
+
+            for optimization in optimizations:
+                if optimization not in self.optimization_map.keys():
+                    logger.e(
+                        logger.Code.INVALID_OPTIMIZATION,
+                        f"The converter doesn't recognise the '{optimization}' optimization.",
+                    )
+
+                # Call the optimization
+                made_changes = self.optimization_map[optimization]()
```

Comment:
FYI, [here](https://github.com/pytorch/executorch/blob/f1ceb6cbdcb8c6e8b8b07d9f0464058a12ceb0d4/backends/arm/tosa_backend.py#L107-L109) is where the graph optimizations are done in PyTorch (Edge IR) before generating a TOSA serialization format.